### ISSUE TITLE
Missing `operator bool()` for marray

### ISSUE THREAD

**Issue by TApplencourt** [KhronosGroup/SYCL-Docs#324]
I think we are don't declare the operator `bool()` for marray.

Indeed it's not present in [4.14.3.1. Math array interface](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_math_array_interface).

And marray doesn't follow the [4.5.2. Common reference semantics](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reference-semantics)

I think that `bool(marray)` should be equal to True if all elements of the array evaluate to `True` and `False` otherwise. 

(found by @nscottnichols)

**Comment by nliber** [comment]
Is there a use case for this?

I'm wary about adding an O(N) (with respect to the number of elements) conversion operator.

**Comment by nscottnichols** [comment]
I cannot think of a good use case for this that is not covered by the `any()` and `all()` functions. I ran into this naively trying something like `result = c ? b : a` (which is covered by `select()`). It might be worthwhile to add a note about undefined behavior.

**Comment by TApplencourt** [comment]
Agree, maybe we just want to be sure that `any()` and `all()` are valid for `marray`. Not sure if this is the case. 
Indeed [all take an vigeninteger](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_relational_functions), and [vigeninteger](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_description_of_the_built_in_types_available_for_sycl_host_and_device)  is only `char{n}, vigeninteger, vugeninteger` and not `marray`.

**Comment by nscottnichols** [comment]
In [Table 182](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.relational.functions.marray), `all` and `any` take `sigeninteger` and `migeninteger`, but again `booln` is not contained in `migeninteger`, which would be useful and well defined here. There are probably other places in the spec where `booln` can be added where `migeninteger` is valid.
